developers/contributing/designs/in-repo-test-config.md

```diff
@@ -0,0 +1,44 @@
+# In-repo test configuration
+
+## Problem Statement
+
+Historically, Zuul uses a configuration file to define which jobs run against which projects.  Each defined job can contain a snippet of shell code to run any arbitrary command or script on the test node to invoke specific test actions against the change under test.  Jobs can also rely on macros to invoke known test suite entry points (ie, tox).
+
+In order to add new jobs to a project being tested by BonnyCI, the Zuul configuration needs to be updated on the Zuul server by BonnyCI admins.  Users of BonnyCI would need to file issues or tickets to get the main Zuul config updated to suit their needs.  This would be a slow process that does not scale, and would make it difficult for new users of BonnyCI to experiment and iterate quickly as they migrate their projects to using BonnyCI.
+
+Other tools such as Travis allow project repositories to carry a .travis.yml file which defines the entire test matrix for the project.  Similar functionality in Zuul is in the works but will not be available until Zuul v3.  Until then, we should provide a flexible way for project maintainers to freely define tests that run against changes.
+
+Additionally, the way in which BonnyCI's Zuul collects test output is dependent on hard-coded config on the server-side.  We need to provide users who are defining their own custom test execution with a standard way of exporting test output so that it may be collected and stored on a log server for post-run viewing.
+
+## Overview
+
+A simple shell interface is provided to users that allows them to store a shell script at a known location in the project repo.  This script is responsible for running the tests shipped within that repository against the change under test. Users are free to define exactly what that script executes so long as it exits with a return code of  0 for a successful test run. Any other return code will be considered a failed run.  Information will be provided to the script, via command-line arguments and environment variables, that describe which stage of testing is running the script (gate, check, etc) and where logs may be stored for publishing by the BonnyCI infrastructure.
+
+On the Zuul single job template will be defined that contains the necessary builders and publishers for executing the testing script and pubishing the test output and logs.  This job template can be used to add jobs to new projects' pipelines. This will help standardize project configuration and streamline the onboarding of new projects into BonnyCI.
```

Comment:
I think this sentence could be improved.

"The Zuul single job template defines the necessary builders for executing the test script and publishers for reporting the test output and logs."